src/App/ComplexGeoData.cpp

```diff
@@ -386,4 +389,208 @@ char ComplexGeoData::elementType(const char *name) const {
+                    if (id == 0)
+                        continue;
+                    auto sid = Hasher->getID(id);
+                    if(!sid)
+                        ++invalid_count;
+                    else
+                        sids.push_back(sid);
+                    char sep;
+                    iss >> sep;
+                }
+            }
+        }
+        _elementMap->setElementName(IndexedName(reader.getAttribute("value"), types),
+                                    MappedName(reader.getAttribute("key")),
+                                    Tag,
+                                    &sids);
+    }
+    if(invalid_count)
+        FC_ERR("Found " << invalid_count << " invalid string id");
+    reader.readEndElement("ElementMap");
+}
+
+void ComplexGeoData::restoreStream(std::istream &s, std::size_t count) {
+    resetElementMap();
+
+    size_t invalid_count = 0;
+    std::string key,value,sid;
+    bool warned = false;
+
+    const auto & types = getElementTypes();
+    try {
+        for(size_t i=0;i<count;++i) {
+            ElementIDRefs sids;
+            std::size_t scount;
+            if(!(s >> value >> key >> scount))
+                FC_THROWM(Base::RuntimeError,
+                          "Failed to restore element map " << _PersistenceName);
+            sids.reserve(scount);
+            for(std::size_t j=0;j<scount;++j) {
+                long id;
+                if(!(s >> id))
+                    FC_THROWM(Base::RuntimeError,
+                              "Failed to restore element map " << _PersistenceName);
+                if (Hasher) {
+                    auto sid = Hasher->getID(id);
+                    if(!sid)
+                        ++invalid_count;
+                    else
+                        sids.push_back(sid);
+                }
+            }
+            if(scount && !Hasher) {
+                sids.clear();
+                if(!warned) {
+                    warned = true;
+                    FC_ERR("missing hasher");
+                }
+            }
+            _elementMap->setElementName(IndexedName(value.c_str(), types), MappedName(key), Tag, &sids);
+        }
+    } catch (Base::Exception &e) {
+        e.ReportException();
+        _restoreFailed = true;
+        _elementMap.reset();
+    }
+    if(invalid_count)
+        FC_ERR("Found " << invalid_count << " invalid string id");
+}
+
+void ComplexGeoData::SaveDocFile(Base::Writer &writer) const {
+    flushElementMap();
+    if (_elementMap) {
+        writer.Stream() << "BeginElementMap v1\n";
+        _elementMap->save(writer.Stream());
+    }
+}
+
+void ComplexGeoData::RestoreDocFile(Base::Reader &reader) {
+    std::string marker, ver;
+    reader >> marker;
+    if (boost::equals(marker, "BeginElementMap")) {
+        resetElementMap();
+        reader >> ver;
+        if (ver != "v1")
+            FC_WARN("Unknown element map format");
+        else {
+            resetElementMap(std::make_shared<ElementMap>());

```

Comment:
I don't really like that reset takes in a parameter. make_shared should be inside reset i think.

Reply:
That's a bigger refactor that's out of scope for this PR, IMO.